sympy/core/mod.py

```diff
@@ -139,6 +137,18 @@ def doit(p, q):
                 net = prod_mod1*prod_mod
                 return prod_non_mod*cls(net, q)
 
+            if q.is_Integer and q is not S.One:
+                _ = []
+                for i in non_mod_l:
+                    if i.is_Integer:
+                        # and (i % q is not S.Zero)
```

Comment:
If `i%q` is zero then the result will be zero. 

I think you better only do this if the denominator is 1, too. Otherwise `Mul(8,4,evaluate=False)%4` would give 0 instead of 2.

Reply:
I am not getting your point, can you please elaborate it !

Reply:
What does `Mod(4*var('x',integer=True)/var('y',integer=True))%8` give in this branch? It should not give 0.

Reply:
Given expression isn't correct one !!!
but for `Mod(4*var('x',integer=True),var('y',integer=True))%8` it gives `Mod(Mod(4*x, y), 8)`

Reply:
Sorry, what about `Mod(8*var('x',integer=True)/var('y',integer=True), 4)`?

Reply:
It gives `4*Mod(2*x/y, 1)` which is correct answer.